### PR TITLE
Latin qx final

### PR DESCRIPTION
**Pull request by yoad-hordan4**
Context
The change was made mainly to address additional Latin medical abbreviations that weren't supported.  these include formats such as "q4d" - take once every 4 hours, "qwk" - take once weekly

Description
this was done mainly through the integration of a dedicated JSON file with all relevant abbreviations and processing of the frequency tag to allow capitalization and periodization

Changes in the codebase
The changes are the said JSON file, additional Latin testers, functions that deal with frequency for such cases, a dedicated class for processing these frequencies, and recognition of 'tab', 'tabs', 'tablet', and 'tablets' as a single entity (tablet) in the preprocessing stage.

Changes outside the codebase
No changes were made outside the codebase

### CODE REVIEW

File: parsigs/frequency_mapping.json [hunk 1]
```diff
@@ -0,0 +1,222 @@
+{
```
Comment (royashcenazi) [inline parsigs/frequency_mapping.json:1]: rename to something like
`latin_freq_dictionary`

File: parsigs/parse_sig_api.py [hunk 2]
```diff
@@ -61,6 +60,13 @@ class StructuredSig:
 inflect_engine = inflect.engine()
 
 
+def read_json_file():  # add parameters path, name
+    base_dir1 = os.path.dirname(__file__)
+    file_path1 = os.path.join(base_dir1, 'frequency_mapping.json')
```
Comment (royashcenazi) [inline parsigs/parse_sig_api.py:65]: almost never there is a good reason to add a number to a variable name

File: parsigs/parse_sig_api.py [hunk 3]
```diff
@@ -73,14 +79,19 @@ def _create_spell_checker():
...
-latin_frequency_types = {"qd": _Frequency("Day", 1), "bid": _Frequency("Day", 2), "tid": _Frequency("Day", 3),
-                         "qid": _Frequency("Day", 4)}
+@dataclass(frozen=True, eq=True)
+class QXFrequency:
+    frequencyType: str
+    interval: int
+
+
+latin_frequency_types = {"qd": Frequency("Day", 1), "bid": Frequency("Day", 2), "tid": Frequency("Day", 3),
+                         "qid": Frequency("Day", 4), "qw": Frequency("Week", 1), "qwk": Frequency("Week", 1)}
```
Comment (royashcenazi) [inline parsigs/parse_sig_api.py:95]: use ths new file to represent this as well and delete it

File: parsigs/parse_sig_api.py [hunk 4]
```diff
@@ -210,20 +221,26 @@ def _create_structured_sig(model_entities, drug=None, form=None):
         elif label == 'Frequency':
             structured_sig.frequencyType = _get_frequency_type(text)
             interval_text = _get_string_after_keyword(text, "every")
+            latin_json = read_json_file()
+            latin_qx = _get_qx_interval(text, latin_json)
```
Comment (royashcenazi) [inline parsigs/parse_sig_api.py:225]: notice this should come just before using it

File: parsigs/parse_sig_api.py [hunk 5]
```diff
@@ -338,16 +355,33 @@ def _get_amount_from_frequency_tags(frequency):
...
+        return latin_freq.interval
+    else:
+        return 1
+
+
 def _get_latin_frequency(frequency):
     for latin_freq in latin_frequency_types.keys():
-        if latin_freq in frequency:
-            return latin_frequency_types[latin_freq]
+        final_freq = latin_freq.lower().replace('.', '')
```
Comment (royashcenazi) [inline parsigs/parse_sig_api.py:375]: rename to `stripped` or something similar

Comment (royashcenazi) [reply parsigs/parse_sig_api.py:375]: no need tyo lower probably - check

File: parsigs/parse_sig_api.py [hunk 6]
```diff
@@ -338,16 +355,33 @@ def _get_amount_from_frequency_tags(frequency):
...
+        if final_freq in frequency:
```
Comment (royashcenazi) [inline parsigs/parse_sig_api.py:376]: this check is redundant in the best case and buggy in the worst if frequency has `.` in it

File: parsigs/parse_sig_api.py [hunk 7]
```diff
@@ -338,16 +355,33 @@ def _get_amount_from_frequency_tags(frequency):
...
+            return latin_frequency_types[final_freq]
+
+
+def _get_qx_interval(frequency, latin_json):
```
Comment (royashcenazi) [inline parsigs/parse_sig_api.py:380]: read as a dict 
string -> `QXFrequency`

File: parsigs/parse_sig_api.py [hunk 8]
```diff
@@ -61,6 +60,13 @@ class StructuredSig:
 inflect_engine = inflect.engine()
 
 
+def read_json_file():  # add parameters path, name
```
Comment (noamlevi113) [inline parsigs/parse_sig_api.py:63]: this function should not be public change the name to `_read_json_file` or create new module that responsible for files general operations (think about indicative name)

Comment (noamlevi113) [reply parsigs/parse_sig_api.py:63]: this function can read only one json file the value that stores in `file_path1`

Comment (noamlevi113) [reply parsigs/parse_sig_api.py:63]: if you will need in the future to use multiple json files in the program you will need to create the same function again. instead you need to create function with this signiture: `def read_json_file(path: str)-> None:`

File: parsigs/parse_sig_api.py [hunk 9]
```diff
@@ -61,6 +60,13 @@ class StructuredSig:
...
+    base_dir1 = os.path.dirname(__file__)
+    file_path1 = os.path.join(base_dir1, 'frequency_mapping.json')
+    with open(file_path1, 'r') as file1:
+        return json.load(file1)
+
+
 def _create_spell_checker():
     sc = SpellChecker()
     drug_words_frequency = Path(__file__).parent / 'resources/drug_names.txt'
```
Comment (noamlevi113) [inline parsigs/parse_sig_api.py:72]: create `config.py` file - inside should be all program constant values for example fils paths that used across the program.

File: tests/test_parsig_api.py [hunk 10]
```diff
@@ -131,10 +156,10 @@ def test_parse_multiple_instructions2(self):
         self.assertEqual(result, expected)
 
     def test_parse_multiple_instructions3(self):
-        sig = "take two tablets of benadryl every two days and then 1 tablet every week for 1 month and than 1 tablet every month"
+        sig = "take two tablets of benadryl every two days and then 1 tablet q.7.d for 1 month and than 1 tablet every month"
         first_expected = StructuredSig(drug="benadryl", form="tablet", strength=None, frequencyType="Day", interval=2,
```
Comment (noamlevi113) [inline tests/test_parsig_api.py:160]: choose more indicative name - the first of what?

File: parsigs/parse_sig_api.py [hunk 11]
```diff
@@ -73,14 +79,19 @@ def _create_spell_checker():
 
 
 @dataclass(frozen=True, eq=True)
-class _Frequency:
+class Frequency:
```
[the remark this answers is not in the document]
Comment (noamlevi113) [reply parsigs/parse_sig_api.py]: still open - you don't have to do it but at least write response of your decision

File: tests/test_parsig_api.py (hunk 10 above)
Comment (noamlevi113) [reply tests/test_parsig_api.py]: is it fixed? (I did not see the changes in the code)

Comment (yoad-hordan4) [reply tests/test_parsig_api.py]: The naming wasn't done by me I only returned the content to its previous status

File: parsigs/parse_sig_api.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (yoad-hordan4) [reply parsigs/parse_sig_api.py]: This class was always in the parse_sig_api file

File: parsigs/parse_sig_api.py [hunk 12]
```diff
@@ -61,6 +67,16 @@ class StructuredSig:
 inflect_engine = inflect.engine()
 
 
+def _open_latin_type_dict():  # add parameters path, name
```
Comment (royashcenazi) [inline parsigs/parse_sig_api.py:70]: @yoad-hordan4 this file needs to be in the `resources` folder
To get its path after for opening, look at the create spell checker `def`
`Path(__file__).parent / 'resources/$FILE_NAME'`

File: parsigs/parse_sig_api.py [hunk 13]
```diff
@@ -61,6 +67,16 @@ class StructuredSig:
...
+    latin_frequency_base_dir = os.path.dirname(__file__)
+    latin_frequency_file_path = os.path.join(latin_frequency_base_dir, 'latin_frequency.json')
+    with open(latin_frequency_file_path, 'r') as file1:
```
Comment (royashcenazi) [inline parsigs/parse_sig_api.py:73]: rename `file1`

File: parsigs/parse_sig_api.py [hunk 14]
```diff
@@ -338,16 +337,11 @@ def _get_amount_from_frequency_tags(frequency):
...
-    latin_freq = _get_latin_frequency(frequency)
-    if latin_freq:
-        return latin_freq.times
-
-
-def _get_latin_frequency(frequency):
-    for latin_freq in latin_frequency_types.keys():
-        if latin_freq in frequency:
-            return latin_frequency_types[latin_freq]
+def _get_latin_frequency(frequency: str):
```
Comment (royashcenazi) [inline parsigs/parse_sig_api.py:340]: this code can be improved by much, think how - where you want to save the same calculation hapenning over and over again. 
hint - ` stripped_freq = frequency.split()[0].replace('.', '')`

File: tests/test_parsig_api.py [hunk 15]
```diff
@@ -95,6 +93,33 @@ def test_parse_sig_latin2(self):
...
 
+    def test_parse_sig_latin3(self):  # test the latin hour abbreviation q_h (take every _ hours)
+        sig = "1 TAB of BENADRYL q7h times for 2 weeks"
+        expected = StructuredSig(drug="benadryl", form='tablet', strength=None, interval=7, frequencyType="Hour",
+                                 times=1, singleDosageAmount=1.0, periodType="Week", periodAmount=2, takeAsNeeded=False)
+        result = self.sig_parser.parse(sig)[0]
+        self.assertEqual(result, expected)
+
+    def test_parse_sig_latin_with_period(self):
+        sig = "take 4 TABS q.4.d for 4 weeks"
```
Comment (royashcenazi) [inline tests/test_parsig_api.py:104]: Add tests for out of standard latin use-cases (tid, bid)

File: parsigs/parse_sig_api.py (hunk 14 above)
Comment (yoad-hordan4) [reply parsigs/parse_sig_api.py:340]: yes an obvious fix, I moved that line out of the for. That is what you meant?

File: parsigs/parse_sig_api.py [hunk 16]
```diff
@@ -61,6 +67,15 @@ class StructuredSig:
 inflect_engine = inflect.engine()
 
 
+def _open_latin_type_dict():  # add parameters path, name
+    latin_frequency_file_path = Path(__file__).parent / 'resources/latin_frequency.json'
```
Comment (royashcenazi) [inline parsigs/parse_sig_api.py:71]: check why a space is after `/`

File: parsigs/parse_sig_api.py [hunk 17]
```diff
@@ -212,16 +216,13 @@ def _create_structured_sig(model_entities, drug=None, form=None):
...
             else:
-                structured_sig.interval = 1
-
-            times_text = _get_string_until_keyword(text, "times")
-            structured_sig.times = _get_amount_from_frequency_tags(times_text)
-
-            if structured_sig.times is None:
-                structured_sig.times = _get_times_from_latin(text)
-
-            # Default added only if there is a frequency tag in the sig, handles cases such as "Every TIME_UNIT"
```
Comment (royashcenazi) [inline parsigs/parse_sig_api.py:224]: return comment

File: parsigs/parse_sig_api.py [hunk 18]
```diff
@@ -212,16 +216,13 @@ def _create_structured_sig(model_entities, drug=None, form=None):
             interval_text = _get_string_after_keyword(text, "every")
             if len(interval_text) > 0:
                 structured_sig.interval = _get_amount_from_frequency_tags(interval_text)
+            latin_frequency_object = _get_latin_frequency(text)
+            if latin_frequency_object:
+                # updates the _Frequency attributes of latin_frequency_object into the structured_sig object
```
Comment (royashcenazi) [inline parsigs/parse_sig_api.py:221]: delete comment

**Review by royashcenazi — APPROVED** [review]
Great Job :-)

**Pull request merged**